### DISCUSSION TITLE
Module not found: Can't resolve 'fs'

### DISCUSSION THREAD

**Discussion by GodsonAddy** [vercel/next.js#45784] [Help]
Hello guys! I need help over here. I'm building an app with next13, nextauth and sequelize.  I'm getting an error which says
 "Module not found: Can't resolve 'fs'"

![Screenshot (175)](https://user-images.githubusercontent.com/70777404/218108657-75a9ccaf-f023-4cf8-8e3a-6a4a19aaebef.png)

next.config.js

`
/** @type {import('next').NextConfig} */
const nextConfig = {
  experimental: {
    appDir: true,
  },
  reactStrictMode: true,
};

module.exports = nextConfig;
`

`
"next": "13.1.6",
 "next-auth": "^4.19.2",
"pg": "^8.9.0",
 "pg-hstore": "^2.3.4",
"react": "18.2.0",
 "react-dom": "18.2.0",
 "sequelize": "^6.28.0",
"@next-auth/sequelize-adapter": "^1.0.7",
"node": "16.15.1"
`

And in my root folder which has my sequelize database where the issue is been highlighted:

db/models/index.js

`
"use strict";
const fs = require("fs");
const path = require("path");
const Sequelize = require("sequelize");
const process = require("process");
const basename = path.basename(__filename);
const env = process.env.NODE_ENV || "development";
const config = require(__dirname + "/../config/config.json")[env];
const db = {};

const models = process.cwd() + "/db/models/" || __dirname;

let sequelize;

if (config.use_env_variable) {
  sequelize = new Sequelize(process.env[config.use_env_variable], config);
} else {
  sequelize = new Sequelize(
    config.database,
    config.username,
    config.password,
    config
  );
}

fs.readdirSync(__dirname)
  .filter((file) => {
    return (
      file.indexOf(".") !== 0 &&
      file !== basename &&
      file.slice(-3) === ".js" &&
      file.indexOf(".test.js") === -1
    );
  })
  .forEach((file) => {
    //const model = require(path.join(__dirname, file))(sequelize, Sequelize.DataTypes);
    const model = path.join(models, file);
    db[model.name] = model;
  });

Object.keys(db).forEach((modelName) => {
  if (db[modelName].associate) {
    db[modelName].associate(db);
  }
});

db.sequelize = sequelize;
db.Sequelize = Sequelize;

module.exports = db;
`

pages/api/auth/[...nextauth].js

`
import NextAuth from "next-auth";
import SequelizeAdapter from "@next-auth/sequelize-adapter";
import GoogleProvider from "next-auth/providers/google";
import { sequelize } from "../../../db/models";
import User from "../../../db/models/user";

export const authOptions = {
  providers: [
    GoogleProvider({
      clientId: process.env.GOOGLE_CLIENT_ID,
      clientSecret: process.env.GOOGLE_CLIENT_SECRET,
    }),
  ],
  adapter: SequelizeAdapter(sequelize),
  callbacks: {
    async session({ session, token, user }) {
      if (!session) return;
      const useremail = session?.user?.email;
      const data = await User.findOne({
        where: { email: useremail },
      });
      session.user.email = data.email;
      session.user.name = data.name;
      session.user.image = data.image;
      session.user.color = data.color;
      session.user.id = data.id;
      return session;
    },
  },
  debug: process.env.NODE_ENV === "development",
  secret: process.env.NEXTAUTH_SECRET,
};

export default NextAuth(authOptions);

`
app/page.js

`import Navbar from "./navbar";

export default function Home() {
  return (

    <div>
      <Navbar />
    </div>

  );
}
`

app/navbar.js

`
"use client"

import { getServerSession } from "next-auth/next";
import { authOptions } from "../pages/api/auth/[...nextauth]";

export default async function Navbar() {
  const session = await getServerSession(authOptions);

return (
    
    <div>
        {session.user.name}
    </div>
)
}

`

I think the issue is because I'm using getServerSession to get a user's session. When I use useSession, it works but it fetches the user's session via the client side(there's a delay when fetching) instead of via server-side
Please any help will be appreciated

**Comment by icyJoseph** [comment]
Hi,

You are putting a `"use client"` directive inside a server component. When the compiler tries to bundle that up for the client it touches `fs`.

Remove that `"use client"` directive and you'll be fine.

---

To format code it is better to use triple `:

> like this: ```js
 
and then put your code

And then close with triple ` once again:

> like this: ```

So that it can look like:

```js
let world = "world"
const hello = `${world}`
```

**Comment by GodsonAddy** [comment]
Thanks @icyJoseph. So should I use getServerSession() inside a server component to fetch user's session or I should use useSession() ?

**Reply by icyJoseph** [reply]
If it works go for getServerSession and keep it as a Server Component.

Using a client component, is mostly for interactivity. 

This is beta technology so I'd say reading the docs and examples, to form an mental model is crucial right now.

**Comment by MedBader007** [comment]
![IMG-20231024-WA0009.jpg](https://github.com/vercel/next.js/assets/146144405/04b2c756-f673-4c66-927a-2ede193e98c3)



Hello guys sorry for the inconvenience I have a question about a problem in my Nextjs Project and I want some help if possible if you can give me that

it gives me this Problem **Module not found: Can't resolve 'fs' ** 

i'm using version 13.5.5 i will show u my package.json for my dependencies


![IMG-20231024-WA0011.jpg](https://github.com/vercel/next.js/assets/146144405/2a034615-4ec3-4580-b5f7-06cb1649ca7c)

**Reply by Manishbatra4** [reply]
i am face same issue did you got any solution?

**Reply by naman-hl** [reply]
Getting same issue with using latest next Js version 14. Please help some one.

**Reply by icyJoseph** [reply]
You have to understand that this issue is not always the "same" issue, different projects, will walk into it in different ways.

The root cause though, your import structure, when fully resolved, is importing `fs`, or any other Node.js API, into a Browser environment.

So you have to figure out which import might be doing this... 

For example, you might have:

```js
import fs from 'fs';

export const MY_CONSTANT = 'foo'

// do stuff with `fs` and `MY_CONSTANT`
```

And then, this constant is imported into another file, where it is used within a function, and that function is also exported, and so on and so on, until you end up using it in a client component, or in general, a module that is meant to run in the browser only.

The bundler, can sometimes, figure out that maybe you only wanted MY_CONSTANT, but it can't always... so you end up having `fs` in scope.

It can even be a `node_modules` file that does this... for example in the root of this comment, I see that a Navbar component, imports from an `action` file, which imports from `api/auth/[...nextAuth]`, which imports from bcrypt, and so on... Navbar has now, transitively, dependency on bcrypt and its dependencies, and if some module uses `fs` there, you won't be able to pack Navbar for the browser.

**Reply by naman-hl** [reply]
@icyJoseph Thanks for reply man!
I m using ```puppeteer-core``` library to use the web scraper in my next js application. And Definately it is causing issue from some of the node_modules file. bcz I m not using fs module directly anywhere.
Do you have any idea how can we get rid of this error ? As in the latest next js v14 , we want to use server actions most of the time.

**Reply by icyJoseph** [reply]
Hi,

But you don't do any of that work from the client right? I mean, the execution of puppeteer-core APIs happens server side, right?

**Reply by naman-hl** [reply]
not exactly, because we have one button on the client side and on pressing that button we are calling the main puppeteer scraping function which is giving us the specific data. So we can say we are trying this on the client side as well.

**Reply by naman-hl** [reply]
```
import puppeteer from 'puppeteer-core'

async function main(keyword) {
	console.log('Connecting to Scraping Browser...')
	const browser = await puppeteer.connect({
		// BrowserWSEndpoint: environmentServerConfig.SBR_WS_ENDPOINT
		browserWSEndpoint:
			'wss://brd-customer-hl_c76b7e37-zone-scraping_browser:<mypassword>@brd.superproxy.io:9222'
	})
	try {
		console.log('Connected! Navigating to Goodreads search results...')
		const page = await browser.newPage()
		page.setDefaultNavigationTimeout(2 * 60 * 1000)
		await page.goto(`https://www.goodreads.com/search?q=${keyword}`)
		console.log('Navigated! Scraping book titles...')

		const books = await page.$$eval('.bookTitle', (nodes) =>
			nodes.map((node) => node.innerText)
		)

		console.log(`Books with "${keyword}" in the title:`)
		let returnStatement = ''
		books.forEach((book, index) => {
			// console.log(`${index + 1}. ${book}`);
			returnStatement += `${index + 1}. ${book}\n`
		})

		return returnStatement
	} finally {
		await browser.close()
	}
}

export default main
```

And when I m importing this function and using in the store file in next js app, I m getting that "fs" module error.
CC @icyJoseph

**Reply by icyJoseph** [reply]
Ah but that's most likely the issue, as I understand it, you are trying to bundle puppeteer for the browser, but that framework/library can only function, server side I'd guess. 

Could you instead, make a network request to a Next.js API Route/Handler, and there all the puppeteer function, and send the result to the client in the response?
